### PR TITLE
【Hackathon 6th No. 7】Add sinc / sinc_ API to Paddle -part

### PR DESCRIPTION
**Pull request by NKNaN**
### PR Category

User Experience

### PR Types

New features

### Description

Add sinc / sinc_ API to Paddle

### CODE REVIEW

File: python/paddle/tensor/math.py [hunk 1]
```diff
@@ -7652,3 +7652,48 @@ def signbit(x, name=None):
...
+        \right.
+
+    Args:
+        x (Tensor): The input Tensor. Must be one of the following types: float16, float32, float64.
+        name (str, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        out (Tensor): The Tensor of elementwise computed normalized sinc result.
+    """
+    if not isinstance(x, (paddle.Tensor, Variable, paddle.pir.Value)):
```
Comment (zxcd) [inline python/paddle/tensor/math.py:7910]: propose adding dtype judgment, also adding in test_sinc.py.

Comment (NKNaN) [reply python/paddle/tensor/math.py:7910]: added

File: python/paddle/tensor/math.py [hunk 2]
```diff
@@ -7652,3 +7652,69 @@ def signbit(x, name=None):
...
+        raise TypeError(f"x must be tensor type, but got {type(x)}")
+
+    if x.dtype not in [
```
Comment (zxcd) [inline python/paddle/tensor/math.py]: maybe you can use "check_variable_and_dtype", this will make your code look more concise.

Comment (NKNaN) [reply python/paddle/tensor/math.py]: revised

**Comment by NKNaN** [comment]
The CI-Coverage check for python code included the last two line of signbit API accidentally.

File: python/paddle/__init__.py [hunk 3]
```diff
@@ -487,6 +487,8 @@
     signbit,
     sin,
     sin_,
+    sinc,
```
Comment (zxcd) [inline python/paddle/__init__.py:494]: also add it in the list of __all__

File: test/legacy_test/test_sinc.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            x_data = np.random.rand(6).astype('int64')
+            startup_program = paddle.static.Program()
+            main_program = paddle.static.Program()
+            exe = base.Executor(place)
+            with paddle.static.program_guard(main_program, startup_program):
+                x = paddle.static.data(name='x', shape=[6], dtype='int64')
+                res = paddle.sinc(x)
+                static_result = exe.run(feed={'x': x_data}, fetch_list=[res])[0]
+
+    def test_inplace(self):
```
Comment (zxcd) [inline test/legacy_test/test_sinc.py:174]: the type error of inplace API also should be test.

File: python/paddle/__init__.py (hunk 3 above)
Comment (NKNaN) [reply python/paddle/__init__.py:494]: added

File: test/legacy_test/test_sinc.py (hunk 4 above)
Comment (NKNaN) [reply test/legacy_test/test_sinc.py:174]: added

File: python/paddle/__init__.py [hunk 5]
```diff
@@ -803,7 +805,8 @@
     'standard_gamma',
     'sinh',
     'sinh_',
-    'round',
+    'sinc',
+    'sinc_' 'round',
```
Comment (zxcd) [inline python/paddle/__init__.py]: attention code style here.

Comment (NKNaN) [reply python/paddle/__init__.py]: oops..
fixed

**Review by zxcd — APPROVED** [review]
LGTM

File: python/paddle/tensor/math.py [hunk 6]
```diff
@@ -7728,3 +7728,70 @@ def signbit(x, name=None):
...
+    Returns:
+        out (Tensor): The Tensor of elementwise computed normalized sinc result.
+    """
+    if not isinstance(x, (paddle.Tensor, Variable, paddle.pir.Value)):
+        raise TypeError(f"x must be tensor type, but got {type(x)}")
+
+    check_variable_and_dtype(
+        x,
+        "x",
+        [
```
Comment (jeff41404) [inline python/paddle/tensor/math.py:7916]: should we support `bfloat16`?

File: python/paddle/tensor/math.py [hunk 7]
```diff
@@ -7728,3 +7728,70 @@ def signbit(x, name=None):
...
+            'float16',
+            'float32',
+            'float64',
+        ],
+        "sinc",
+    )
+
+    tmp = paddle.where(x != 0, x, paddle.full_like(x, 1.0e-20))
```
Comment (jeff41404) [inline python/paddle/tensor/math.py:7925]: why use `1.0e-20`, if data type of `x` is float16, will there be underflow?

File: python/paddle/tensor/math.py [hunk 8]
```diff
@@ -7728,3 +7728,70 @@ def signbit(x, name=None):
...
+    Inplace version of ``sinc`` API, the output Tensor will be inplaced with input ``x``.
+    Please refer to :ref:`api_paddle_sinc`.
+    """
+    if not isinstance(x, (paddle.Tensor, Variable)):
+        raise TypeError(f"x must be tensor type, but got {type(x)}")
+
+    check_variable_and_dtype(
+        x,
+        "x",
+        [
```
Comment (jeff41404) [inline python/paddle/tensor/math.py:7943]: should we support `bfloat16`?

File: python/paddle/tensor/math.py [hunk 9]
```diff
@@ -7728,3 +7728,70 @@ def signbit(x, name=None):
...
+            'float16',
+            'float32',
+            'float64',
+        ],
+        "sinc_",
+    )
+
+    paddle.where_(x != 0, x, paddle.full_like(x, 1.0e-20))
```
Comment (jeff41404) [inline python/paddle/tensor/math.py:7952]: same issue as above

File: python/paddle/tensor/math.py (hunk 6 above)
Comment (NKNaN) [reply python/paddle/tensor/math.py:7916]: yes.
`bfloat16` added

File: python/paddle/tensor/math.py (hunk 8 above)
Comment (NKNaN) [reply python/paddle/tensor/math.py:7943]: yes.
`bfloat16` added

File: python/paddle/tensor/math.py (hunk 7 above)
Comment (NKNaN) [reply python/paddle/tensor/math.py:7925]: `1.0e-20` is used in numpy.sinc, but there is underflow problem when data type is `float16`.
Fixed the problem using `paddle.isnan` before returning the final value.

File: python/paddle/tensor/math.py (hunk 9 above)
Comment (NKNaN) [reply python/paddle/tensor/math.py:7952]: Fixed the problem using `paddle.isnan` before returning the final value.

File: python/paddle/tensor/math.py (hunk 7 above)
Comment (jeff41404) [reply python/paddle/tensor/math.py:7925]: Can we support backward gradient calculation, especially when x=0?

Comment (jeff41404) [reply python/paddle/tensor/math.py:7925]: shall we add test case for calculating gradient

Comment (NKNaN) [reply python/paddle/tensor/math.py:7925]: > Can we support backward gradient calculation, especially when x=0?

when `x=0`，the gradient should be 0:  

$$( \frac{\pi x}{\sin(\pi x)} )^{\prime} = \pi \frac{\sin(\pi x) - \pi x \cos(\pi x)}{\sin^2(\pi x)}$$

by LHopital's rule,

$$ \lim_{x \rightarrow 0} \frac{\sin(\pi x) - \pi x \cos(\pi x)}{\sin^2(\pi x)} = \lim_{x \rightarrow 0} \frac{\pi \cos(\pi x) - [\pi \cos(\pi x) - \pi^2 x \sin(\pi x) ]}{2 \pi \sin(\pi x) \cos(\pi x)} =  \lim_{x \rightarrow 0} \frac{\pi x}{2 \pi cos(\pi x)} = 0$$

The torch example:
```python
x = torch.rand(2,3)
x[0,0] = 0.0
x.requires_grad = True
s = torch.sinc(x).sum()
s.backward()
x.grad
# tensor([[ 0.0000, -1.2824, -1.0302],
#        [-1.3687, -1.0435, -1.1842]])
```

Comment (NKNaN) [reply python/paddle/tensor/math.py:7925]: > shall we add test case for calculating gradient

I will add it

Comment (NKNaN) [reply python/paddle/tensor/math.py:7925]: > Can we support backward gradient calculation, especially when x=0?

when `x=0`，the gradient should be 0:  

$$( \frac{\sin(\pi x)}{\pi x} )^{\prime} = \pi \frac{\pi x \cos(\pi x) - \sin(\pi x)}{\pi^2 x^2}$$

by LHopital's rule,

$$ \lim_{x \rightarrow 0} \frac{\pi x \cos(\pi x) - \sin(\pi x)}{\pi^2 x^2} = \lim_{x \rightarrow 0} \frac{ [\pi \cos(\pi x) - \pi^2 x \sin(\pi x) ] - \pi \cos(\pi x)}{2 \pi^2 x} =  \lim_{x \rightarrow 0} -\frac{\sin(\pi x)}{2} = 0$$

The torch example:
```python
x = torch.rand(2,3)
x[0,0] = 0.0
x.requires_grad = True
s = torch.sinc(x).sum()
s.backward()
x.grad
# tensor([[ 0.0000, -1.2824, -1.0302],
#        [-1.3687, -1.0435, -1.1842]])
```
The current implementation using `paddle.where` can return gradient value 0 if x = 0

Comment (NKNaN) [reply python/paddle/tensor/math.py:7925]: > Can we support backward gradient calculation, especially when x=0?

Should inplace operation `sinc_` also support gradient calculation?

**Comment by luotao1** [comment]
image
@NKNaN 看下这里的覆盖率，如果本地单测有覆盖到，请截图贴上。

**Comment by NKNaN** [comment]
@NKNaN 看下这里的覆盖率，如果本地单测有覆盖到，请截图贴上。

这个是 isreal 的内容，不知道为啥这三行也会算到 coverage 检测里面

**Comment by luotao1** [comment]
> 这个是 isreal 的内容，不知道为啥这三行也会算到 coverage 检测里面

收到，已经豁免了

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by sunzhongkai588 — APPROVED** [review]
LGTM
